src/input/key-state.c

```diff
@@ -69,20 +69,32 @@ key_state_nr_pressed_sent_keycodes(void)
 }
 
 void
-key_state_set_pressed(uint32_t keycode, bool ispressed)
+key_state_set_pressed(uint32_t keycode, bool is_pressed, bool is_modifier)
 {
-	if (ispressed) {
+	if (is_pressed) {
 		add_key(&pressed, keycode);
+		if (is_modifier) {
+			add_key(&pressed_mods, keycode);
+		}
 	} else {
 		remove_key(&pressed, keycode);
+		remove_key(&pressed_mods, keycode);
 	}
 }
 
 void
-key_state_store_pressed_keys_as_bound(void)
+key_state_store_pressed_key_as_bound(uint32_t keycode)
 {
-	memcpy(bound.keys, pressed.keys, MAX_PRESSED_KEYS * sizeof(uint32_t));
-	bound.nr_keys = pressed.nr_keys;
+	add_key(&bound, keycode);
```

Comment:
In the …bound() function do we need to nullify the array before adding?

bound.nr_keys = 0

Reply:
I think it doesn't matter one way or the other. If there are still pressed/held modifiers from a previous keybinding (that haven't been released yet) then we will add them a second time, but IIRC add_key() has a check to avoid adding a duplicate key-code.

Reply:
Actually, thinking about this more, I think we do not want to clear/nullify the bound array before adding. There may be a non-modifier key still pressed from a previous keybinding (if you pressed two in quick succession) and that should still be considered "bound".

Reply:
When I made the comment, I had forgotten that there is a `key_state_bound_key_remove()`, so thought we just kept adding :smile:

You make a good point. Agree.